csharp/src/Ice/InputStream-Class.cs

```diff
@@ -311,90 +359,217 @@ private AnyClass ReadInstance(int index)
+                        if (SkipSlice(allTypeIds[i]))
+                        {
+                            Debug.Assert(i == skipCount - 1);
+                            break;
+                        }
+                        else
+                        {
+                            ReadNextSliceHeaderIntoCurrent();
+                            ReadIndirectionTableIntoCurrent();
+                        }
+                    }
                 }
-                else if (SkipSlice()) // Slice off what we don't understand.
+                else if (formalTypeId != null)
                 {
-                    instance = new UnknownSlicedClass(this);
+                    // received null and formalTypeId is not null, apply formal type optimization.
+                    if (Communicator.FindClassFactory(formalTypeId) is IClassFactory factory)
+                    {
+                        instance = factory.Create();
+                        _instanceMap.Add(instance);
+                        ReadIndirectionTableIntoCurrent();
+                        // Nothing to skip
+                    }
+                    else
+                    {
+                        throw new ArgumentException($"cannot find class factory for `{formalTypeId}'",
+                            nameof(formalTypeId));
+                    }
+                }
+                else
+                {
+                    throw new InvalidDataException(
+                        @$"cannot read a class instance with no type ID; did you forget to specify the {
+                            nameof(formalTypeId)}?");
                 }
             }
-            while (instance == null);
 
-            Pop(previousCurrent);
+            instance.Read(this);
+
+            _current = previousCurrent;
             --_classGraphDepth;
             return instance;
         }
 
-        // Reads a slice header into _current.
-        // Returns the type ID of that slice. Null means it's a slice in compact format without a type ID, or a slice
-        // with a compact ID we could not resolve.
-        private string? ReadSliceHeaderIntoCurrent()
+        /// <summary>Reads the header of the current slice into _current; this method is used when the current slice
+        /// is not the first (most derived) slice.</summary>
+        private void ReadNextSliceHeaderIntoCurrent()
+        {
+            if (OldEncoding)
+            {
+                // With the 1.1 encoding, each slice header in sliced format contains a type ID - we read it and
+                // ignore it.
+                _ = ReadSliceHeaderIntoCurrent11();
+            }
+            else
+            {
+                Debug.Assert(_current.InstanceType != InstanceType.None);
+
+                _current.SliceFlags = (EncodingDefinitions.SliceFlags)ReadByte();
+
+                Debug.Assert((EncodingDefinitions.TypeIdKind)
```

Comment:
We could add an extension method to the SliceFlags enum to extract the TypeIdKind and make this a little more readable.

Reply:
Fixed as suggested.